### PR TITLE
[local_proxy] update api for pg_session_jwt

### PR DESCRIPTION
**Pull request by conradludgate**
pg_session_jwt now:
1. Sets the JWK in a PGU_BACKEND session guc, no longer in the init() function.
2. JWK no longer needs the kid.

### CODE REVIEW

File: proxy/src/serverless/local_conn_pool.rs [hunk 1]
```diff
@@ -475,20 +457,39 @@ impl LocalClient<tokio_postgres::Client> {
...
+        info!(pid, jti = inner.jti, "user session state init");
 
         Ok(())
     }
 }
 
-fn sign_jwt(sk: &SigningKey, header: String, payload: String) -> String {
-    let header = Base64UrlUnpadded::encode_string(header.as_bytes());
-    let payload = Base64UrlUnpadded::encode_string(payload.as_bytes());
+fn sign_jwt(sk: &SigningKey, payload: &[u8]) -> String {
```
Comment (conradludgate) [inline proxy/src/serverless/local_conn_pool.rs:466]: I had a bit of time so I thought I'd try make this work with a single alloc only :)

File: proxy/src/serverless/backend.rs [hunk 2]
```diff
@@ -276,45 +277,46 @@ impl PoolingBackend {
...
-            .query(
-                "select auth.init($1, $2);",
-                &[
-                    &kid as &(dyn ToSql + Sync),
-                    &tokio_postgres::types::Json(jwk),
-                ],
-            )
-            .await?;
+            // initiates the auth session
+            if let Err(e) = client.query("select auth.init();", &[]).await {
```
[the remark this answers is not in the document]
Comment (conradludgate) [reply proxy/src/serverless/backend.rs:313]: it is not necessary. in fact this query() block can only accept a single query due to the way the postgres protocol works!

File: proxy/src/serverless/backend.rs [hunk 3]
```diff
@@ -276,45 +277,46 @@ impl PoolingBackend {
...
+        let key = SigningKey::random(&mut OsRng);
+        let jwk = p256::PublicKey::from(key.verifying_key()).to_jwk();
+        let jwk = serde_json::to_string(&jwk).expect("serializing jwk to json should not fail");
+
         let config = node_info
             .config
             .user(&conn_info.user_info.user)
-            .dbname(&conn_info.dbname);
+            .dbname(&conn_info.dbname)
+            .options(&format!("-c pg_session_jwt.jwk={jwk}"));
```
[the remark this answers is not in the document]
Comment (conradludgate) [reply proxy/src/serverless/backend.rs:288]: yeah it's weird :/ I've tested it and the `-c` is necessary. So strange.

File: proxy/src/serverless/backend.rs [hunk 4]
```diff
@@ -276,45 +277,46 @@ impl PoolingBackend {
...
             node_info.aux.clone(),
         );
 
-        let kid = handle.get_client().get_process_id() as i64;
-        let jwk = p256::PublicKey::from(handle.key().verifying_key()).to_jwk();
-
-        debug!(kid, ?jwk, "setting up backend session state");
+        {
+            let (client, mut discard) = handle.inner();
+            debug!(?pid, "setting up backend session state");
```
[the remark this answers is not in the document]
Comment (conradludgate) [reply proxy/src/serverless/backend.rs:310]: not necessary. can probably remove it. `kid=pid` and I just updated it to match.

File: proxy/src/serverless/local_conn_pool.rs (hunk 1 above)
Comment (conradludgate) [reply proxy/src/serverless/local_conn_pool.rs:490]: Yeah, probably should do that 😅

File: proxy/src/serverless/local_conn_pool.rs [hunk 5]
```diff
@@ -542,3 +559,19 @@ impl<C: ClientInnerExt> Drop for LocalClient<C> {
...
+
+    use super::resign_jwt;
+
+    #[test]
+    fn jwt_token_snapshot() {
+        let key = SigningKey::from_bytes(&[1; 32].into()).unwrap();
+        let data = br#"{"foo":"bar", "jti": "foo\nbar", "nested": {"jti": "tricky nesting"}}"#;
+
+        let jwt = resign_jwt(&key, data, 2).unwrap();
+        assert_eq!(jwt, "e30.eyJmb28iOiJiYXIiLCJqdGkiOjIsIm5lc3RlZCI6eyJqdGkiOiAidHJpY2t5IG5lc3RpbmcifX0.qZnHxDfNigd7eegbHtcjK-l3Gv-zQtx6-6x-uP1IBfwS4Z4Jepc3T3RfRaIbPpWOjeVBrhLjB3jir5_LtSVPXQ");
```
[the remark this answers is not in the document]
Comment (conradludgate) [reply proxy/src/serverless/local_conn_pool.rs:575]: As for how I validated the test, I copied the token into <https://jwt.io/> and manually validated that the token met my expectations

[the remark this answers is not in the document]
Comment (cloneable) [reply proxy/src/serverless/local_conn_pool.rs:575]: I tried exactly that, but couldn't figure out how to use the 32 0x01s for signing. Would be good to have at least a little comment describing this. Was hoping for a cli tool command one could run, but linking to jwt.io is okay, too.

File: compute/Dockerfile.compute-node [hunk 6]
```diff
@@ -925,8 +925,8 @@ ARG PG_VERSION
 RUN case "${PG_VERSION}" in "v17") \
     echo "pg_session_jwt does not yet have a release that supports pg17" && exit 0;; \
     esac && \
-    wget https://github.com/neondatabase/pg_session_jwt/archive/ff0a72440e8ff584dab24b3f9b7c00c56c660b8e.tar.gz -O pg_session_jwt.tar.gz && \
-    echo "1fbb2b5a339263bcf6daa847fad8bccbc0b451cea6a62e6d3bf232b0087f05cb pg_session_jwt.tar.gz" | sha256sum --check && \
+    wget https://github.com/neondatabase/pg_session_jwt/archive/b30d7df196d0cb816365e565a7579f9c78221206.tar.gz -O pg_session_jwt.tar.gz && \
```
Comment (mrl5) [inline compute/Dockerfile.compute-node:928]: one more fix was needed, please change commit hash to `5aee2625af38213650e1a07ae038fdc427250ee4`

File: compute/Dockerfile.compute-node [hunk 7]
```diff
@@ -925,8 +925,8 @@ ARG PG_VERSION
...
+    echo "ea9a666af4bcdd84e8363f56d0b730849bfcc94db4f01713d579859a46435818 pg_session_jwt.tar.gz" | sha256sum --check && \
```
Comment (mrl5) [inline compute/Dockerfile.compute-node:929]: one more fix was needed, please update sha256sum accordingly
(on my end it's `5d91b10bc1347d36cffc456cb87bec25047935d6503dc652ca046f04760828e7` now)

**Review by mrl5 — CHANGES_REQUESTED** [review]

File: proxy/src/serverless/local_conn_pool.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (conradludgate) [reply proxy/src/serverless/local_conn_pool.rs]: Updated. Can you test the process?

**Review by mrl5 — APPROVED** [review]

**Review by cloneable — APPROVED** [review]

File: proxy/src/serverless/local_conn_pool.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (cloneable) [reply proxy/src/serverless/local_conn_pool.rs]: Yes, works! I guess going from the 32 0x01 signing key seed to a key is impl specific and can't be tested easily externally.

**Pull request merged**